fx quant: fix bug with fusion on a module used by multiple nodes

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68178

Summary:

Before this PR, the FX quantization fusion logic incorrectly fused
modules which had multiple nodes using them.  For example,

```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(1, 1, 1)
        self.bn = torch.nn.BatchNorm2d(1)

    def forward(self, x):
        x = self.conv(x)
        x = self.conv(x)
        x = self.bn(x)
        return x
```

the second conv-bn would be fused, which would change the meaning of
the first call to the conv.

This PR changes the pattern matching logic to disable matching
for any nodes which are not the last node in a pattern and which
refer to modules which have multiple module_uses.

Note that this definition of "module_use" is different from the "use" in the
FX framework.  In FX, a use is referring to which nodes use the output
of a node.  "module_use" refers to how many FX nodes use a module
instance.

Test Plan:

```
python test/test_quantization.py TestFuseFx.test_fusion_turns_off_for_multi_use_intermediate_nodes
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32357549](https://our.internmc.facebook.com/intern/diff/D32357549)

vkuzo committed:
fx quant: fix bug with fusion on a module used by multiple nodes

Summary:

Before this PR, the FX quantization fusion logic incorrectly fused
modules which had multiple nodes using them.  For example,

```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(1, 1, 1)
        self.bn = torch.nn.BatchNorm2d(1)

    def forward(self, x):
        x = self.conv(x)
        x = self.conv(x)
        x = self.bn(x)
        return x
```

the second conv-bn would be fused, which would change the meaning of
the first call to the conv.

This PR changes the pattern matching logic to disable matching
for any nodes which are not the last node in a pattern and which
refer to modules which have multiple module_uses.

Note that this definition of "module_use" is different from the "use" in the
FX framework.  In FX, a use is referring to which nodes use the output
of a node.  "module_use" refers to how many FX nodes use a module
instance.

Test Plan:

```
python test/test_quantization.py TestFuseFx.test_fusion_turns_off_for_multi_use_intermediate_nodes
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32357549).

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -2,23 +2,29 @@
+        calculate_module_name_to_num_node_users(model.graph)
     for n in model.graph.nodes:
-        if not is_match(modules, n, pattern):
+        if not is_match(modules, n, pattern, module_name_to_num_node_users):
```
does it work if we just set `max_uses = 1` here?

vkuzo replied:
sorry, not sure I understand the question yet, could you provide some more context?

`max_uses` is an FX concept, this PR is adding checking for "number of nodes which use this module instance" which is not a concept that exists in FX.

jerryzh168 replied:
oh right, sorry I mixed the use of Node and use of module instance

jerryzh168 commented on torch/ao/quantization/fx/match_utils.py:
```diff
@@ -49,16 +75,30 @@ def is_match(modules, node, pattern, max_uses=sys.maxsize):
 
+    if (
+        # This is True if we are not at the last node in the pattern
+        max_uses < sys.maxsize and
```
could you explain a little bit more about what does this mean and why do we need this check?

vkuzo replied:
you are right, this check is incorrect, if the last node in the pattern is using a module which is used by other nodes we also cannot fuse without further analysis.  Will remove this.

vkuzo committed:
Update on "fx quant: fix bug with fusion on a module used by multiple nodes"


Summary:

Before this PR, the FX quantization fusion logic incorrectly fused
modules which had multiple nodes using them.  For example,

```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(1, 1, 1)
        self.bn = torch.nn.BatchNorm2d(1)

    def forward(self, x):
        x = self.conv(x)
        x = self.conv(x)
        x = self.bn(x)
        return x
```

the second conv-bn would be fused, which would change the meaning of
the first call to the conv.

This PR changes the pattern matching logic to disable matching
for any nodes which are not the last node in a pattern and which
refer to modules which have multiple module_uses.

Note that this definition of "module_use" is different from the "use" in the
FX framework.  In FX, a use is referring to which nodes use the output
of a node.  "module_use" refers to how many FX nodes use a module
instance.

Test Plan:

```
python test/test_quantization.py TestFuseFx.test_fusion_turns_off_for_multi_use_intermediate_nodes
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32357549](https://our.internmc.facebook.com/intern/diff/D32357549)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32357549).

vkuzo committed:
Update on "fx quant: fix bug with fusion on a module used by multiple nodes"


Summary:

Before this PR, the FX quantization fusion logic incorrectly fused
modules which had multiple nodes using them.  For example,

```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(1, 1, 1)
        self.bn = torch.nn.BatchNorm2d(1)

    def forward(self, x):
        x = self.conv(x)
        x = self.conv(x)
        x = self.bn(x)
        return x
```

the second conv-bn would be fused, which would change the meaning of
the first call to the conv.

This PR changes the pattern matching logic to disable matching
for any nodes which are not the last node in a pattern and which
refer to modules which have multiple module_uses.

Note that this definition of "module_use" is different from the "use" in the
FX framework.  In FX, a use is referring to which nodes use the output
of a node.  "module_use" refers to how many FX nodes use a module
instance.

Test Plan:

```
python test/test_quantization.py TestFuseFx.test_fusion_turns_off_for_multi_use_intermediate_nodes
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32357549](https://our.internmc.facebook.com/intern/diff/D32357549)

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: fix bug with fusion on a module used by multiple nodes"


Summary:

Before this PR, the FX quantization fusion logic incorrectly fused
modules which had multiple nodes using them.  For example,

```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(1, 1, 1)
        self.bn = torch.nn.BatchNorm2d(1)

    def forward(self, x):
        x = self.conv(x)
        x = self.conv(x)
        x = self.bn(x)
        return x
```

the second conv-bn would be fused, which would change the meaning of
the first call to the conv.

This PR changes the pattern matching logic to disable matching
for any nodes which are not the last node in a pattern and which
refer to modules which have multiple module_uses.

Note that this definition of "module_use" is different from the "use" in the
FX framework.  In FX, a use is referring to which nodes use the output
of a node.  "module_use" refers to how many FX nodes use a module
instance.

Test Plan:

```
python test/test_quantization.py TestFuseFx.test_fusion_turns_off_for_multi_use_intermediate_nodes
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32357549](https://our.internmc.facebook.com/intern/diff/D32357549)

[ghstack-poisoned]

vkuzo commented:
@jerryzh168 , please check out if the modified test cases are what you expect

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32357549).

vkuzo committed:
Update on "fx quant: fix bug with fusion on a module used by multiple nodes"


Summary:

Before this PR, the FX quantization fusion logic incorrectly fused
modules which had multiple nodes using them.  For example,

```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(1, 1, 1)
        self.bn = torch.nn.BatchNorm2d(1)

    def forward(self, x):
        x = self.conv(x)
        x = self.conv(x)
        x = self.bn(x)
        return x
```

the second conv-bn would be fused, which would change the meaning of
the first call to the conv.

This PR changes the pattern matching logic to disable matching
for any nodes which are not the last node in a pattern and which
refer to modules which have multiple module_uses.

Note that this definition of "module_use" is different from the "use" in the
FX framework.  In FX, a use is referring to which nodes use the output
of a node.  "module_use" refers to how many FX nodes use a module
instance.

Test Plan:

```
python test/test_quantization.py TestFuseFx.test_fusion_turns_off_for_multi_use_intermediate_nodes
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32357549](https://our.internmc.facebook.com/intern/diff/D32357549)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32357549).

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -445,6 +446,53 @@ def forward(self, x):
         })
         self.checkGraphModuleNodes(m, expected_node=ns.call_module(MyConvReLU))
 
+    def test_fusion_turns_off_for_multi_use_intermediate_nodes(self):
```
can we also have a test for multiple use of relu module? like conv - relu?

jerryzh168 reviewed (approved):
lg, one comment inline for test

vkuzo committed:
Update on "fx quant: fix bug with fusion on a module used by multiple nodes"


Summary:

Before this PR, the FX quantization fusion logic incorrectly fused
modules which had multiple nodes using them.  For example,

```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(1, 1, 1)
        self.bn = torch.nn.BatchNorm2d(1)

    def forward(self, x):
        x = self.conv(x)
        x = self.conv(x)
        x = self.bn(x)
        return x
```

the second conv-bn would be fused, which would change the meaning of
the first call to the conv.

This PR changes the pattern matching logic to disable matching
for any nodes which are not the last node in a pattern and which
refer to modules which have multiple module_uses.

Note that this definition of "module_use" is different from the "use" in the
FX framework.  In FX, a use is referring to which nodes use the output
of a node.  "module_use" refers to how many FX nodes use a module
instance.

Test Plan:

```
python test/test_quantization.py TestFuseFx.test_fusion_turns_off_for_multi_use_intermediate_nodes
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32357549](https://our.internmc.facebook.com/intern/diff/D32357549)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32357549).

vkuzo committed:
Update on "fx quant: fix bug with fusion on a module used by multiple nodes"


Summary:

Before this PR, the FX quantization fusion logic incorrectly fused
modules which had multiple nodes using them.  For example,

```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(1, 1, 1)
        self.bn = torch.nn.BatchNorm2d(1)

    def forward(self, x):
        x = self.conv(x)
        x = self.conv(x)
        x = self.bn(x)
        return x
```

the second conv-bn would be fused, which would change the meaning of
the first call to the conv.

This PR changes the pattern matching logic to disable matching
for any nodes which are not the last node in a pattern and which
refer to modules which have multiple module_uses.

Note that this definition of "module_use" is different from the "use" in the
FX framework.  In FX, a use is referring to which nodes use the output
of a node.  "module_use" refers to how many FX nodes use a module
instance.

Test Plan:

```
python test/test_quantization.py TestFuseFx.test_fusion_turns_off_for_multi_use_intermediate_nodes
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32357549](https://our.internmc.facebook.com/intern/diff/D32357549)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32357549).

jerryzh168 commented on torch/ao/quantization/fx/match_utils.py:
```diff
@@ -22,17 +23,49 @@
+# nodes. Module types not in this list will not be matched if they are used by
+# multiple 'call_module' nodes.
+MODULE_TYPES_SAFE_TO_USE_BY_MULTIPLE_NODES = set([
+    torch.nn.ReLU,
```
I remember Kimish reported another case, there is a Lienar - relu pattern, and then then people access the bias in linear, should we add Linear/Conv here as well?

jerryzh168 replied:
Oh sorry, this is safe to use, sounds good

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -469,6 +470,105 @@ def forward(self, x):
+            expected_node_list=expected_nodes,
+            expected_node_occurrence=expected_occurrence)
+
+    def test_fusion_works_for_multi_use_intermediate_relu_nodes(self):
```
there is one more test case, reported by Kimish, could you add that as well (probably need to change the matching code to address this as well), something like

```
class M(torch.nn.Module):
    def forward(self, x):
           x = self.linear(x)
           bias = self.linear.bias
           ...
```
I can take it as well if you don't have time
